pkg/plugin/kubernetes.go

```diff
@@ -0,0 +1,218 @@
+	}
+
+	name := defaultVeleroDeploymentName
+	if opts.veleroDeploymentName != "" {
+		name = opts.veleroDeploymentName
+	}
+
+	existingDeployment, err := clientset.AppsV1().Deployments(os.Getenv("VELERO_NAMESPACE")).Get(context.TODO(), name, metav1.GetOptions{})
+	if kuberneteserrors.IsNotFound(err) {
+		return nil, errors.Wrap(err, "velero deployment not found")
+	} else if err != nil {
+		return nil, errors.Wrap(err, "failed to get velero deployment")
+	}
+
+	return existingDeployment, nil
+}
+
+func ensureDeploymentHasVolume(deployment *appsv1.Deployment, volumeSpec *corev1.Volume, volumeMountSpec *corev1.VolumeMount, opts localVolumeObjectStoreOpts) error {
+
+	// If the volume name is the same, but the path is different, we should fix the path in place
+	if exists, idx := podHasDuplicateVolumeName(&deployment.Spec.Template.Spec, volumeSpec); exists {
+		deployment.Spec.Template.Spec.Volumes[idx] = *volumeSpec
+	} else {
+		deployment.Spec.Template.Spec.Volumes = append(deployment.Spec.Template.Spec.Volumes, *volumeSpec)
+
+		// TODO (dans): user configuration for velero container name
+		veleroContainer := getContainerByName(deployment, "velero")
+		if veleroContainer == nil {
+			return errors.New("velero container not found")
+		}
+		veleroContainer.VolumeMounts = append(veleroContainer.VolumeMounts, *volumeMountSpec)
+
+		// Add the POD_IP for servering the signed URLs
+		if !containerHasEnvVar(veleroContainer, "POD_IP") {
+			veleroContainer.Env = append(veleroContainer.Env, corev1.EnvVar{
+				Name: "POD_IP",
+				ValueFrom: &corev1.EnvVarSource{
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: "status.podIP",
+					},
+				},
+			})
+		}
+	}
+
+	// TODO (dans): make sure that the MOUNT_POINT env exists, even if the container is already there.
+	fileServerContainer := getContainerByName(deployment, fileServerContainerName)
+
+	fileServerImage := defaultFileServerContainerImage
+	if opts.fileserverImage != "" {
+		fileServerImage = opts.fileserverImage
+	}
+
+	if fileServerContainer == nil {
+		fileServerContainer = &corev1.Container{
+			Name:  fileServerContainerName,
+			Image: fileServerImage,
+			Env: []corev1.EnvVar{
+				{
+					Name:  "MOUNT_POINT",
+					Value: getRoot(),
+				},
+			},
+		}
+		deployment.Spec.Template.Spec.Containers = append(deployment.Spec.Template.Spec.Containers, *fileServerContainer)
+	}
+	fileServerContainer.VolumeMounts = append(fileServerContainer.VolumeMounts, *volumeMountSpec)
+
+	clientset, err := k8sutil.GetClientset()
+	if err != nil {
+		return errors.Wrap(err, "unable to get kubernetes clientset")
+	}
+
+	_, err = clientset.AppsV1().Deployments(os.Getenv("VELERO_NAMESPACE")).Update(context.TODO(), deployment, metav1.UpdateOptions{})
+	if err != nil {
+		return errors.Wrap(err, "unable to update velero deployment")
+	}
+
+	return nil
+}
+
+func peakDameonset(opts localVolumeObjectStoreOpts) (*appsv1.DaemonSet, error) {
```

Comment:
since this is returning the daemonset pointer, should this be "getResticDaemonset" to be more descriptive?

Reply:
Yeah, I was trying to find a way to differentiate the above which return an error if it doesn't exist and this method that returns `nil`, but the naming is confusing. Will update.